[Ng Chi Sern] iP

imposterble opened the pull request:
## Botto
> “Bottom line is, if you do not use it or need it, it’s clutter, and it needs to go.” - Charisse Ward

**Botto frees your mind of having to remember things you need to do. It's**
* text-based
* easy to learn
* ~~FAST~~ **EXTREMELY FAST** to use

**All you need to do is,**
1. Download the source code [here](https://github.com/ngchisern/ip/releases/tag/v0.1).
2. Run the application. 
3. Add your tasks.
4. Let it manage your tasks for you :dizzy:

**Features**
- [x] List all the tasks
- [x] Add/delete a task 
- [x] Mark a task as done 
- [x] Search tasks using keywords 
- [ ] Reminders (coming soon)

**Sneak peek of my `main` method:**
```java
public static void main(String[] args) {
        new Botto("data/botto.txt").run();
}
```

Commit:
Add Gradle support

imposterble committed:
complete level 1 (Greet, Echo, Exit)

imposterble committed:
complete Level 2 (Add, List)

imposterble committed:
complete Level 3 (Mark as Done)

imposterble committed:
complete Level 4 (ToDos, Events, Deadlines)

imposterble committed:
complete A-TextUiTesting

imposterble committed:
Complete Level 5 (Handle Errors)

imposterble committed:
Compelete Level 6 (Delete)

imposterble committed:
use Java enum for task type

imposterble committed:
Add save feature

Save the tasks in the hard disk automatically whenever the task list changes. Load the data from the hard disk when Duke starts up.

imposterble committed:
Add Dates and Times

Teach Botto how to understand dates and times. For example, if the command is deadline return book /by 2/12/2019 6:00 pm, Botto should understand 2/12/2019 6:00 pm as 2nd of December 2019, 6pm, instead of treating it as just a String.

imposterble committed:
Merge branch 'branch-Level-7'

imposterble committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Botto.java

Resolution: combine both of the features

imposterble committed:
Make the code more OOP

Add Ui, Storage, Parser, TaskList and Command classes

imposterble committed:
Divide classes into packages

Create botto, botto.command, botto.util and botto.task packages

imposterble committed:
Add JUnit tests

* Add test for all public methods in Todo class
* Add test for all public methods in Deadline class
* Add test for all public methods in Event class
* Add test for all public methods in Parser class
* Add test for all public methods in TaskList class
* Add test for some public methods in Ui class

imposterble committed:
Package the App as a JAR file

imposterble committed:
Add JavaDoc comments

Add JavaDoc comments to all classes and methods

imposterble committed:
Tweak the code to comply with a coding standard

imposterble committed:
Implement find feature

Give users a way to find a task by searching for a keyword.

imposterble committed:
Merge branch 'branch-A-JavaDoc'

imposterble committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/botto/Botto.java
#	src/main/java/botto/BottoException.java
#	src/main/java/botto/command/Command.java
#	src/main/java/botto/task/Deadline.java
#	src/main/java/botto/task/Event.java
#	src/main/java/botto/util/Parser.java
#	src/main/java/botto/util/Storage.java
#	src/main/java/botto/util/TaskList.java

imposterble committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/botto/task/Task.java
#	src/main/java/botto/util/Parser.java
#	src/main/java/botto/util/Ui.java
#	text-ui-test/sources.txt

imposterble committed:
Merge branch 'add-gradle-support' into branch-A-Gradle

imposterble committed:
Set up gradle to build and run Botto using gradle.

imposterble committed:
Merge branch 'branch-A-Gradle'

imposterble committed:
Use CheckStyle and fix some styling issues

imposterble committed:
Merge branch 'branch-A-CheckStyle'

skythefire commented on src/main/java/botto/util/Storage.java:
```diff
@@ -0,0 +1,109 @@
+package botto.util;
```
Good usage of packaging different classes into different packages!

skythefire commented on src/main/java/botto/util/Ui.java:
```diff
@@ -0,0 +1,125 @@
+ * This class deals with the Botto bot's interactions with the user
+ */
+public class Ui {
+    private static final String BOT_NAME = "Botto";
+    private static final String INDENTATION = "    ";
```
I like how you uppercased your constant names :D

skythefire commented on src/main/java/botto/command/AddEventCommand.java:
```diff
@@ -0,0 +1,62 @@
+/**
+ * Command for adding an event
+ */
+public class AddEventCommand implements Command {
```
I like how you had a Command interface where classes with similar behavior implement it!

skythefire commented on src/main/java/botto/util/Parser.java:
```diff
@@ -0,0 +1,67 @@
+    public static Command parse(String fullCommand) throws BottoException {
+        String command = findCommand(fullCommand);
+
+        switch (command) {
+        case "list":
+            return new ShowListCommand();
+        case "done":
+            return new MarkDoneCommand(fullCommand);
+        case "todo":
+            return new AddToDoCommand(fullCommand);
+        case "deadline":
+            return new AddDeadlineCommand(fullCommand);
+        case "event":
+            return new AddEventCommand(fullCommand);
+        case "delete":
+            return new DeleteCommand(fullCommand);
+        case "bye":
+            return new ExitCommand();
+        case "find":
+            return new FindCommand(fullCommand);
+        default:
+            throw new BottoException("☹ OOPS!!! I'm sorry, but I don't know what that means :-(");
```
I like your usage of the switch case statement here!

skythefire reviewed (commented):
LGTM! Good effort and all the best with the remaining iP tasks :")

imposterble committed:
Add a GUI to Botto

 Use the JavaFX technology to implement the GUI.

imposterble committed:
Merge branch 'branch-Level-10'

jeremy-7007 commented on src/main/java/botto/command/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,61 @@
+ * Command for adding a deadline
+ */
+public class AddDeadlineCommand implements Command {
+    private String command;
```
Instead of passing a command String to each command object, you can consider passing a Task object, and leave the parsing logic below to your Parser class! That way, you can also merge AddDeadlineCommand, AddTodoCommand and AddEventCommand into a single AddCommand.

jeremy-7007 commented on src/test/java/botto/util/ParserTest.java:
```diff
@@ -0,0 +1,54 @@
+public class ParserTest {
+
+    @Test
+    public void parse_allCommands_success() throws BottoException {
```
Although it is more tedious, it might be better to split each command type into its own test function. But that's not a big deal.

jeremy-7007 reviewed (approved):
Suggested an optional structural change to the project regarding the add-command-classes. Otherwise, no coding violation that I can find. LGTM!

imposterble committed:
Add shadow plugin to create a new JAR file

imposterble committed:
Use assertions in Parser, Storage and TaskList classes

These classes assume that the task list, command and file path passed to their constructors and methods are not null.

Assertions can be used to check the state a code expects before it starts running or state after it finishes running.

Let's,
* Add assertions to handle the cases which these parameters are null.

imposterble committed:
Improve code quality

The program was not written according to the guidelines taught in the [latest lecture](https://nus-cs2103-ay2122s1.github.io/website/schedule/week5/topics.html#w5-4-code-quality).

The code quality's guidelines help in improving readability and avoiding unsafe practices.

Let's,
* use refactoring methods to improve the code quality

imposterble committed:
Merge pull request #1 from ngchisern/branch-A-Assertions

Use assertions in Parser, Storage and TaskList classes

imposterble committed:
Merge branch 'master' into branch-A-CodeQuality

* master:
  Use assertions in Parser, Storage and TaskList classes

imposterble committed:
Merge pull request #2 from ngchisern/branch-A-CodeQuality

Improve code quality

imposterble committed:
Create gradle.yml

imposterble committed:
Merge branch 'branch-A-CI'

* branch-A-CI:
  Create gradle.yml

imposterble committed:
Delete irrelevant test cases

imposterble committed:
Attempt to solve unsuccessful CI workflow run

imposterble committed:
Attempt to solve unsuccessful CI workflow run (part 2)

imposterble committed:
Add sorting functionality

Without sorting, the organisation of the tasks can be very messy when they are printed

Let's sort the tasks by
* type of tasks
* target time of the tasks

imposterble committed:
Merge branch 'branch-C-Sort'

* branch-C-Sort:
  Add sorting functionality

imposterble committed:
Enable assertions in build.gradle

imposterble committed:
Improve the GUI

imposterble committed:
Add the storage file to git ignore

imposterble committed:
Delete storage file

imposterble committed:
Add a brief user guide and Ui.png

imposterble committed:
Add help command

imposterble committed:
Set theme jekyll-theme-architect

imposterble committed:
Improve method/class header comments according to the format specified in the coding standard

imposterble committed:
Merge branches 'master' and 'master' of https://github.com/ngchisern/ip

* 'master' of https://github.com/ngchisern/ip:
  Set theme jekyll-theme-architect

* 'master' of https://github.com/ngchisern/ip:
  Set theme jekyll-theme-architect

imposterble committed:
Improve code quality

imposterble committed:
Add more Junit Tests and improve Find&Help command's header comment

imposterble committed:
Update botto version

imposterble committed:
Fix storage loading issue and standardise command format for datetime

imposterble committed:
Update a hyperlink (in the user guide) to the latest Botto's release version

imposterble committed:
Improve method header comments

* In method header comments, the first sentence should start in the form Returns ..., Sends ..., Adds ... (not Return or Returnning etc.)

imposterble committed:
Update README.md
